File: questoesSlidesHaskell/IsPrime.hs
isPrime:: Int -> Bool
isPrime 1 = False
isPrime 2 = True
isPrime num = checkForDivisors num 2

checkForDivisors:: Int -> Int -> Bool
checkForDivisors num current =
  if current + 1 == num
  then isNotDivisible num current
  else (isNotDivisible num current) && (checkForDivisors num (current + 1))

isNotDivisible:: Int -> Int -> Bool
isNotDivisible dividendo divisor = (mod dividendo divisor) /= 0

main:: IO ()
main =
  do
    num <- readLn :: IO Int
    putStrLn(show(isPrime num))

File: questoesSlidesProlog/isPrime.pl
/*
um inteiro X é primo se (X mod 2 != 0) && (X mod 3 != 0) && (X mod 4 != 0)
&& ... (X mod (X -1) != 0)
*/
checkForDivisors(X, X).
checkForDivisors(X, Y):-
  X mod Y =\= 0,
  Z is Y + 1,
  checkForDivisors(X, Z).

isPrime(1).
isPrime(2).
isPrime(X):-
  checkForDivisors(X, 2).
  
main:-
  read(X),
  isPrime(X).